snapshot-labs/sx-monorepo
# feat: vote modal with reason

### Summary



Closes: #386

This PR moves the vote UI inside a vote modal, similar to v1, and allow attaching a reason to a vote.

### Features

- new vote modal: voting for a choice will open a Vote modal, where user can see a summary of its vote and voting power, before submitting the vote. This step will add an extra guard layer, where we can prevent users without sufficient voting power from voting.
- vote with reason: a new `reason` field has been added in the vote modal, where the voter can attach a reason to his vote.
- refactoring the `getVotingPower` logic into a store and composable, similar to how we handle `Space`, enabling caching and avoid code duplicate

### How to test

1. Try to vote from inside a proposal page, or a proposal list (space proposals/home feed)
2. It should open the vote modal if logged in, else will open the account modal
3. The vote modal should show your selected choice, as well as your voting power for the current proposal
4. The `CONFIRM` button should be disabled if you don't have sufficient voting power to vote
5. You can attach an optional reason to your vote
6. Vote is limited to 1000 chars, and should disable the form on error
7. Clicking on the CONFIRM should vote like before if no reason, and will add a `metadataUri` (onchain) or `reason` (offchain) field if set
8. It should show a warning message and disable the submit button when the vp fetching has failed or on insufficient vp
9. On the create proposal page and vote modal, it should show a `RETRY` button to refetch the voting power, in case of fetching error
10. Voting power should be cached, you should not see the loading indicator when getting the same vp again on same space and block, and should show the vp immediately. 
11. Voting power should refresh on wallet change/logout

### TODO

- [x] Confirm reason works offchain 
- [ ] Confirm reason works on EVM testnet (waiting for thegraph fix on the sepolia graph)
- [ ] Confirm reason works on EVM mainnet
- [x] Confirm reason works on Starknet testnet
- [ ] Confirm reason works on Starknet mainnet (signature error, waiting for https://github.com/snapshot-labs/sx-monorepo/pull/464 as potential fix)
- [x] Fix/Update tests

### Future

- The rest of #386 (showing tx status and share buttons) will be implemented in #443, as this PR is already pretty stuffed.
